src/wg/format/png/reader.d

```diff
@@ -0,0 +1,926 @@
+                    output[writePos++] = input[x+2];
+                    if (input[x..x + 2] == transparency[0..3]) output[writePos++] = 0;
+                    else output[writePos++] = 0xff;
+                    x += 3;
+                }
+                output = output[newLineBytes..$];
+                input = input[lineBytes..$];
+            }
+        }
+
+        result.rowPitch = newLineBytes;
+        return true;
+    }
+
+    private bool setupFormat() nothrow @nogc
+    {
+        static immutable string[] formats = [
+            "",
+            "l_1",
+            "l_2",
+            "l_4",
+            "l",
+            "l_16",
+            "rgb",
+            "rgb_16_16_16",
+
+            "",
+            "la_1_1",
+            "la_2_2",
+            "la_4_4",
+            "la",
+            "la_16_16",
+            "rgba",
+            "rgba_16_16_16_16",
+        ];
+
+        int intFormat;
+        import core.bitop: bsf;
+        final switch (colorType)
+        {
+            case PngColorType.grayscale:
+                intFormat = bitDepth.bsf() + 1;
+                if (transLength > 0)
+                {
+                    if (!moveGrayscaleTransparencyToAlpha()) return false;
+                    intFormat |= 8;
+                }
+                break;
+            case PngColorType.grayscaleWithAlpha:
+                intFormat = 8 | (bitDepth.bsf() + 1);
+                break;
+            case PngColorType.paletteColor:
+                intFormat = 6;
+                if (!depalettize()) return false;
+                if (transLength > 0) intFormat |= 8;
+                break;
+            case PngColorType.rgbColor:
+                intFormat = bitDepth == 8 ? 6 : 7;
+                if (transLength > 0) 
+                {
+                    if (!moveRgbTransparencyToAlpha()) return false;
+                    intFormat |= 8;
+                }
+                break;
+            case PngColorType.rgbaColor:
+                intFormat = bitDepth == 8 ? 14 : 15;
+                break;
+        }
+
+        switch (intFormat) {
+            case     1, 2, 3, 4, 5: result.bitsPerBlock = bitDepth; break;
+            case              6, 7: result.bitsPerBlock = cast(ubyte)(3 * bitDepth); break;
+            case 9, 10, 11, 12, 13: result.bitsPerBlock = cast(ubyte)(2 * bitDepth); break;
+            case            14, 15: result.bitsPerBlock = cast(ubyte)(4 * bitDepth); break;
+            default:                result.bitsPerBlock = 0; break;
+        }
+
+        auto strFormat = formats[intFormat];
+
+        if (!isSRGB)
+        {
+            if (chromaticities.isSet)
+            {
+                // TODO: Read chromaticities and try to find standard RGB color space that mathes.
+                // If found and it is not sRGB add its name to strFormat using allocator.
+                // If not found add string in the format "@{wX, wY}R{rX, rY}G{gX, gY}B{bX, bY}".
```

Comment:
It should be `R{rX, rY}G{gX, gY}B{bX, bY}@{wX, wY}`, and you can omit the `@` whitepoint if it is `D65`, which it probably is.

Reply:
Don't forget you still need to add parsing of only two values :)

Reply:
that works fine...?

Reply:
Also, you should use `Wx, Wy`, because `xy` is different than `XY` (the case totally matters) :)

Reply:
...assuming they are `xyY` and not `XYZ`? If there's only 2 values, then it will be `xyY`

Reply:
Sorry, parsing from... text?
`wg.color.xyz : parseXYZ` can be used like `parseXYZ!xyY(text)` That will work with only 2 values.

Reply:
Oh, yes. I missed that. Sorry.